glean-core/src/upload/mod.rs

```diff
@@ -0,0 +1,357 @@
+        // As long as there are pings in the queue this will
+        // keep trying to create a `PingRequest` from it.
+        while ping.is_some() {
+            // Safe unwrap, we state that ping "is_some" before entering the loop
+            let (uuid, path) = ping.unwrap();
+            let request = self.process_ping_file(uuid, path);
+            if request.is_some() {
+                return request;
+            }
+            ping = self.get_next_ping();
+        }
+        None
+    }
+
+    /// Process the next ping, if there are any in the queue.
+    ///
+    /// Will try **once** to re-process the pending_pings folder
+    /// in case the queue is empty.
+    fn get_next_ping(&self) -> Option<(Uuid, PathBuf)> {
+        let mut attempts = 0;
+        loop {
+            if let Some(uuid) = self.queue.process() {
+                // Create the ping file path from the UUID
+                let path = self.pending_pings_dir.join(uuid.to_string());
+                return Some((uuid, path));
+            }
+
+            if let Err(e) = self.process_pings_dir() {
+                log::error!("Error processing pings directory! {}", e);
+            }
+
+            // Will only attempt once to process the
+            // pending pings directory before giving up.
+            if attempts > 0 {
+                break;
+            } else {
+                attempts += 1;
+            }
+        }
+        None
+    }
+
+    /// Process and add all valid ping files to queue.
+    ///
+    /// Any files that don't match the UUID regex will be deleted
+    /// to prevent files from polluting the pings directory.
+    ///
+    /// Any directories will also be deleted. There should only be
+    /// files inside the `pending_pings` directory.
+    fn process_pings_dir(&self) -> Result<()> {
+        for entry in self.pending_pings_dir.read_dir()? {
+            if let Ok(entry) = entry {
+                let path = entry.path();
+
+                if path.is_dir() {
+                    match remove_dir_all(&path) {
+                        Err(e) => log::error!(
+                            "Error deleting folder inside pending_pings dir {}. {}",
+                            path.display(),
+                            e
+                        ),
+                        _ => {
+                            log::warn!("Deleted folder inside pending_pings dir {}", path.display())
+                        }
+                    };
+                    continue;
+                }
+
+                if let Some(file_name) = get_file_name_as_str(&path) {
+                    if !PING_FILE_NAME_PATTERN.is_match(file_name) {
+                        log::warn!("Pattern mismatch {}", path.display());
+                        delete_file(&path);
+                        continue;
+                    }
+
+                    // Safe unwrap, already verified if valid UUID
+                    self.queue.enqueue(Uuid::parse_str(file_name).unwrap());
```

Comment:
If we have pending pings, at startup, we would probably want to send them in a specific order (oldest first!), like we do on desktop. This would make the behaviour consistent once we jump to project FOG.

Note that this doesn't need to happen as part of this PR, but please file a bug and make sure the current design allows for that.

Reply:
I am not sure this is what you mean, but the queue always will return oldest first.

I could maybe loop the folder in a way that the oldest files are enqueued first, thus processed first, if that is what you mean.

Reply:
> I could maybe loop the folder in a way that the oldest files are enqueued first, thus processed first, if that is what you mean.

Yes, that's what I mean, sorry if I was unclear :-)